src/target.rs

```diff
@@ -138,42 +137,136 @@ fn find_newest(path: &str) -> Result<File, Error> {
+            .path()
+            .join("output-file")
+            .to_str()
+            .unwrap()
+            .to_owned();
+
+        match oc {
+            OutputCase::None => (),
+            OutputCase::Missing => args.push(output_path.clone()),
+            OutputCase::Some => {
+                std::fs::File::create_new(&output_path)
+                    .unwrap()
+                    .set_modified(now)
+                    .unwrap();
+                args.push(output_path.clone());
+            }
+        }
+
+        // Setup the input...
+        args.push(":".into());
+        let input_path = tempdir
+            .path()
+            .join("input-file")
+            .to_str()
+            .unwrap()
+            .to_owned();
+
+        match ic {
+            InputCase::None => (),
+            ic @ _ => {
+                std::fs::File::create_new(&input_path)
+                    .unwrap()
+                    .set_modified(ic.permute_time(now))
+                    .unwrap();
+                args.push(input_path);
+            }
+        }
+
+        args.push("--".into());
+        args.push("touch".into());
+        args.push(output_path.into());
+
+        args
+    }
+
+    fn check_build(args: &Vec<String>, expected_needs_build: bool) {
+        let target = Target::parse(args.clone()).unwrap();
+        assert_eq!(target.needs_rebuild, expected_needs_build);
+        assert_eq!(target.should_run_command(), expected_needs_build);
+
+        let expected_code = 0;
+        let actual_code = target.run_command().unwrap();
+        assert_eq!(actual_code, expected_code);
+    }
+
+    #[tracing_test::traced_test]
     #[test]
-    fn test_target() {
-        let tmp_dir = tempdir().unwrap();
-
-        let src = tmp_dir.path().join("a.c");
-        let dest = tmp_dir.path().join("a.out");
-
-        std::fs::write(&src, "int main() { return 0; }").unwrap();
-
-        let target = Target::parse(
-            [
-                "target/debug/mk",
-                dest.to_str().unwrap(),
-                ":",
-                src.to_str().unwrap(),
-                "--",
-                "cc",
-                "-o",
-                dest.to_str().unwrap(),
-                src.to_str().unwrap(),
-            ]
-            .iter()
-            .map(|s| s.to_string())
-            .collect(),
-        );
-        let target = match target {
-            Err(e) => panic!("{}", e),
-            Ok(t) => t,
-        };
-        assert!(target.needs_rebuild);
-        assert!(target.should_run_command());
-        assert_eq!(target.run_command().unwrap(), 0);
-        assert!(std::fs::metadata(&dest).is_ok());
+    fn test_needs_rebuild() {
+        for case in all_test_cases() {
+            let tempdir = tempdir().unwrap();
+            let args = setup_test_case(&tempdir, case.0, case.1);
+
+            let expected_needs_build = match case {
+                (_, OutputCase::None | OutputCase::Missing) => true,
```

Comment:
Something seems off about building when there's no output specified…

Reply:
That shouldn't be a valid case, if it is it's a bug. It should build if there are no _inputs_ specified though.

Reply:
`mk : input-file -- touch xyx` ?

Reply:
There's nothing for it to compare timestamps with, so it's identical to `touch xyz`. I guess it's not really an issue, it's just pointless.

Reply:
But _should_ it be? I mean, should it run `touch xyz` when:

- There's no inputs? (Yes, very explicitly)
- There's no outputs? (It does … but it seems weird)
- There's a newer input than all the outputs? (Yes! _"executing command if any input is newer than all outputs"_)
- If the inputs are all either the same age or older than the output (No)

I suspect it feels weird to me because:

1. My mental model is "any input is newer than all outputs." Something like: `inputs.any(|i| i > outputs.max())`
1. But, if we have `max`, why not express the whole condition in similar terms: `inputs.min() > outputs.max()`
1. What is a `min` or `max` for an empty set? That is…
  1. `[].min()` for `inputs` must be `MAXTIME` because we always want to build when there's no inputs
  2. … so then is `[].max()` for `outputs` also `MAXTIME`? If so, then we _shouldn't_ build when there are no outputs?
1. Or perhaps we never have empty sets?
  - `[MAXTIME, ...inputs].min() > [EPOCH, ...outputs].max()` will trigger builds if there's no inputs `||` if there's no outputs.
1. `MAXTIME` is weird. Why not `now`?
  - Because `[NOW, ...inputs].min() > [EPOCH, ...outputs].max()` won't trigger a build if the there's no inputs and the newest output is newer than now.
  - … frankly, that seems fine _too_?

Anyway, the behavior at the extremes feels weird no matter what way we cut it. Why even accept `mk` with empty outputs? That feels like a bad legacy of `make`. tbh, even accepting with no inputs feels bad too; it's shell, just run the command!

Reply:
That's what I said to begin with...

> That shouldn't be a valid case, if it is it's a bug.

`make` without a target aren't valid, so this shouldn't be.

Reply:
OK, maybe I'll improve the parser and data model?

Reply:
In the same vein: why accept cases without _inputs_?

Reply:
> OK, maybe I'll improve the parser and data model?

I think it's fine as-is

Reply:
> In the same vein: why accept cases without _inputs_?

Because make does